### ISSUE TITLE
DA queue crashing and not clearing disk queue files properly

### ISSUE THREAD

**Issue by dpkkumar01** [rsyslog/rsyslog#3311]
I am reading log data and sending to kafka using DA queue to store my data during the kafka connectivity issues. I have configured the queue size for testing the disk writing and recovery. 

I have made kafka connectivity down and started pushing the traffic. Data has been saved in In-memory and copied to disk after reaching high Watermark level. I have brought up the connectivity and started receiving the message at Kafka.  Initially i have more queue files in the queue directory then reduced to 2 files(mentioned below) which was not cleared from the queue directory even after message delivery.

**-rw------- 1 root root 1435194 Nov 20 20:13 stats_ruleset.00000001
-rw------- 1 root root     599 Nov 20 20:13 stats_ruleset.qi**

Again i have made kafka connectivity down and started pushing the load. Here once high Watermark level reached rsyslog is crashing and going down. I am unable to bring up rsyslog till i clearing the above two queue files.

Below is the log for your reference.
[rsyslog_debug.zip](https://github.com/rsyslog/rsyslog/files/2600323/rsyslog_debug.zip)

**Rsyslog version --> 8.37.0**

**Below is my configuration:**

module(load="imfile" Mode="inotify")
module(load="mmnormalize")
module(load="omkafka")

template (name="MTFW_CDR" type="list") {
     constant(value="{\"ReportGeneratedTime\":\"")
     property(name="$!field1")
     constant(value=" +05:30")
     constant(value="\",\"Delivered Time\":\"")
     property(name="$!field3")
     constant(value=" +05:30")
     constant(value="\",\"Record Type\":\"")
     property(name="$!field5")
     constant(value="\",\"Delivery Status\":\"")
     property(name="$!field17")
     constant(value="\",\"Failure Reason\":\"")
     property(name="$!field18")
     constant(value="\",\"Failure Reason Type\":\"")
     property(name="$!field19")
     constant(value="\",\"Service Centre\":\"")
     property(name="$!field53")
     constant(value="\",\"Vendor\":\"TELEDNA\",\"Location\":\"NOIDA\",\"IPAddress\":\"172.16.23.28\",\"Product\":\"MTFW\"}")
}

input(type="imfile"
     #File="/home/sms/SMSC-Sec-2.0.7.0_MT/cdrs/DNA-AegisGw-CDR_PREPAID*.log"
      File="/home/prem/rsyslog/cdrs/DNA-AegisGw-CDR_PREPAID*"
      Tag="cdr: "
      PersistStateInterval="5000"
      deleteStateOnFileDelete="off"
      ruleset="MTFW_CDR")

ruleset(name="MTFW_CDR") {
     action(type="mmnormalize" rulebase="/etc/MTFW_CDR.rb")
     action(
            action.resumeRetryCount="-1"
            action.resumeInterval="5"
            action.reportSuspension="on"
            action.reportSuspensionContinuation="on"
            action.errorfile="/var/log/kafka_error.log"
            broker=["172.16.20.15:9092"]
            template="MTFW_CDR"
            type="omkafka"
            topic="mtfw_cdr"
            partitions.auto="on"
            errorFile="/var/log/kafka_error.log"
            confParam=[
                       "socket.timeout.ms=1000",
                       "socket.keepalive.enable=true",
                       "queue.buffering.max.messages=2000000",
                       "queue.buffering.max.ms=1000",
                       "compression.codec=gzip"
                       ]
                       queue.type="LinkedList"
                       #queue.workerThreads="1"
                       #queue.dequeueBatchSize="100"
                       queue.highWatermark="1000"
                       #queue.timeoutshutdown="60000"
                       #queue.timeoutEnqueue="60000"
                       #queue.timeoutActionCompletion="60000"
                       #queue.timeoutworkerthreadshutdown="-1"
                       queue.lowWatermark="700"
                       queue.spoolDirectory="/var/run/rsyslog/queues"
                       queue.fileName="stats_ruleset"
                       queue.maxDiskSpace="5g"
                       queue.size="20000"
                       queue.saveOnShutdown="on"
                       queue.CheckpointInterval="1"
                       queue.syncqueuefiles="on"
                       queue.maxFileSize="3m"
           )stop
}

**Comment by davidelang** [comment]
On Tue, 20 Nov 2018, dpkkumar01 wrote:

> Initially i have more queue files in the queue directory then reduced 
> to 2 files(mentioned below) which was not cleared from the queue directory 
> even after message delivery.

This is working as intended, the final files are not deleted until rsyslog is 
shutdown, because a new log could arrive at any time that needs to be written to 
the file, and to avoid the race condition of adding a new log to files that are 
in the middle of being deleted, it doesn't try.

David Lang

**Comment by davidelang** [comment]
On Tue, 20 Nov 2018, dpkkumar01 wrote:

> **-rw------- 1 root root 1435194 Nov 20 20:13 stats_ruleset.00000001
> -rw------- 1 root root     599 Nov 20 20:13 stats_ruleset.qi**
>
> Again i have made kafka connectivity down and started pushing the load. Here once high Watermark level reached rsyslog is crashing and going down. I am unable to bring up rsyslog till i clearing the above two queue files.
>
> Below is the log for your reference.
> [rsyslog_debug.zip](https://github.com/rsyslog/rsyslog/files/2600323/rsyslog_debug.zip)

Is this debug log of a startup with the files still there?

what version of rsyslog are you running?

David Lang

**Comment by dpkkumar01** [comment]
Yes, Above debug log for startup with the file. I have shared this log using rsyslog version 8.37.

Later i have tested with 8.39, I am facing the same problem even there.

**Comment by dpkkumar01** [comment]
@davidelang , Is that shared logs helps to debug this issue??.

If it is enough, Please check and update

**Comment by dpkkumar01** [comment]
@davidelang ,  Kindly check and update on this issue.

**Comment by dpkkumar01** [comment]
@davidelang , Can you please update on this issue

**Comment by davidelang** [comment]
This is something that I can't fix and don't know enough to dive into. I know enough to identify what info is needed for @rgerhards and others to troubleshoot.

**Comment by dpkkumar01** [comment]
@rgerhards , Can you please check this issue and update.

**Comment by rgerhards** [comment]
@dpkkumar01 very short on time, but: do you mean rsyslog aborts? If so, please provide valgrind log of an abort.

**Comment by dpkkumar01** [comment]
@rgerhards , Sorry for the delay.

[valgrind_log.txt](https://github.com/rsyslog/rsyslog/files/2697642/valgrind_log.txt)
[rsyslog_log.zip](https://github.com/rsyslog/rsyslog/files/2697646/rsyslog_log.zip)
Please find the attached logs for your reference.

Kindly do the needful.

**Comment by dpkkumar01** [comment]
@rgerhards , Can you please check this issue.   Let me know if you need any more information on the same.

**Comment by dpkkumar01** [comment]
Hi @rgerhards ,

We are facing this issue in one of our live setup.
I was shared the required logs for debugging this issue. Can you please check .

**Comment by rgerhards** [comment]
sry, had overlooked the valgrind log. Will have a look now.

**Comment by rgerhards** [comment]
mmmhhhh... seems to be a template problem and trigger seems to be that kafka goes down. Unfortunately, the debug log does not provide sufficient info to nail the cause.

Would you be able to build a (special) version yourself and let it run and provide debug log?

**Comment by dpkkumar01** [comment]
Ya sure. I can do that. Let me know what i have to do for collecting these information??

**Comment by rgerhards** [comment]
great! I first need to add extra instrumentation to the code. I'll try today, but can't promise. If you don't hear back by Friday, pls ping me.

**Comment by dpkkumar01** [comment]
@rgerhards , Kindly do the needful..
